### PR TITLE
fix autoremove voucher when changing carrier

### PR DESCRIPTION
**Pull request by tomlev**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Fix auto-remove carrier-specific voucher when changing carrier
| Type?         | bug fix
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | http://forge.prestashop.com/browse/BOOM-3307 http://forge.prestashop.com/browse/BOOM-5218 http://forge.prestashop.com/browse/BOOM-5219
| How to test?  | Add a carrier-specific voucher. Check voucher is automatically removed when changing carrier. Test the same behaviour with empty-code voucher

#### Important guidelines

* Make sure [your local branch is up to date](https://help.github.com/articles/syncing-a-fork/) before commiting your changes!
* Your code MUST respect [our Coding Standards](http://doc.prestashop.com/display/PS16/Coding+Standards) (for code written in PHP, JavaScript, HTML/CSS/Smarty/Twig, SQL)!

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/9142)

### CODE REVIEW

File: tests/Unit/Core/Cart/Calculation/Carrier/RuleWhenChangingCarrierTest.php [hunk 1]
```diff
@@ -0,0 +1,174 @@
+<?php
+/**
+ * 2007-2017 PrestaShop
```
Comment (PierreRambaud) [inline tests/Unit/Core/Cart/Calculation/Carrier/RuleWhenChangingCarrierTest.php]: Wrong year!

Comment (tomlev) [reply tests/Unit/Core/Cart/Calculation/Carrier/RuleWhenChangingCarrierTest.php]: damn

File: tests/Unit/Core/Cart/AbstractCartTest.php [hunk 2]
```diff
@@ -299,22 +299,30 @@ protected function insertCartRule($cartRuleFixtureId, $cartRuleData)
 
     protected function addCartRulesToCart(array $cartRuleFixtureIds)
     {
+        $allAdded = true;
         foreach ($cartRuleFixtureIds as $cartRuleFixtureId) {
             $cartRule = $this->getCartRuleFromFixtureId($cartRuleFixtureId);
-            if ($cartRule !== null) {
+            if ($cartRule === null) {
+                $allAdded = false;
```
Comment (PierreRambaud) [inline tests/Unit/Core/Cart/AbstractCartTest.php:395]: Maybe break when $allAdded is set to false or does it need to add others?

File: tests/Unit/Core/Cart/Calculation/Carrier/RuleWhenChangingCarrierTest.php [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        $cartRules = $this->cart->getCartRules();
+        $this->assertCount(1, $cartRules);
+
+        $this->setCartCarrierFromFixtureId(1);
+
+        $cartRules = $this->cart->getCartRules();
+        $this->assertCount(0, $cartRules);
+    }
+
+    public function testCarrierSpecificCartRuleWithoutCodeSetAndUnset()
```
Comment (mickaelandrieu) [inline tests/Unit/Core/Cart/Calculation/Carrier/RuleWhenChangingCarrierTest.php:156]: I love tests <3

Comment (tomlev) [reply tests/Unit/Core/Cart/Calculation/Carrier/RuleWhenChangingCarrierTest.php:156]: ahah the more explicit, the more maintainable !

File: tests/Unit/Core/Cart/AbstractCartTest.php (hunk 2 above)
Comment (tomlev) [reply tests/Unit/Core/Cart/AbstractCartTest.php:395]: I added a comment : here we just silent not-added cart rules. But some tests have to know the global result

**Review by PierreRambaud — DISMISSED** [review]

**Comment by marionf** [comment]
Need to do some others tests before merge

**Comment by marionf** [comment]
Need a rebase when #9170 will be merged

**Comment by eternoendless** [comment]
@tomlev Could you please rebase onto 1.7.4.x?

**Comment by tomlev** [comment]
rebase does not work... new PR here https://github.com/PrestaShop/PrestaShop/pull/9172

**Comment by tomlev** [comment]
closed as duplicate of https://github.com/PrestaShop/PrestaShop/pull/9172

**Pull request closed without merging**